radis/lbl/base.py

```diff
@@ -2142,50 +2322,101 @@ def calc_populations_noneq(
+                    dfQrot_dict = dict(list(zip(dfQrot.index, dfQrot.Qrot)))
+
+                    dg = df.loc[idx]
+
+                    # Add lower state Qrot
+                    dg_sorted = dg.set_index(["viblvl_l"], inplace=False)
+                    df.loc[idx, "Qrotl"] = dg_sorted.index.map(dfQrot_dict.get).values
+                    # Add upper state energy
+                    dg_sorted = dg.set_index(["viblvl_u"], inplace=False)
+                    df.loc[idx, "Qrotu"] = dg_sorted.index.map(dfQrot_dict.get).values
+
+                    # ... note: the .map() version is about 3x faster than
+                    # ... dg.groupby('viblvl_l').apply(lambda x: dfQrot_dict[x.name])
 
-                # Get partition function for all lines
+                    if radis.DEBUG_MODE:
+                        assert (df.loc[idx, "id"] == id).all()
+                        assert (df.loc[idx, "iso"] == iso).all()
+
+            else:
+                dgb = df.groupby(by=["iso"])
+                id = df.attrs["id"]
                 molecule = get_molecule(id)
-                state = self.input.state
-                parsum = self.get_partition_function_calculator(molecule, iso, state)
-                Q, Qvib, dfQrot = parsum.at_noneq(
-                    Tvib,
-                    Trot,
-                    vib_distribution=vib_distribution,
-                    rot_distribution=rot_distribution,
-                    overpopulation=overpopulation,
-                    returnQvibQrot=True,
-                    update_populations=self.misc.export_populations,
-                )
 
-                # ... make sure PartitionFunction above is calculated with the same
-                # ... temperatures, rovibrational distributions and overpopulations
-                # ... as the populations of active levels (somewhere below)
-                df.at[idx, "Qvib"] = Qvib
-                df.at[idx, "Q"] = Q
+                for (iso), idx in dgb.indices.items():
+
+                    # Get partition function for all lines
+                    state = self.input.state
+                    parsum = self.get_partition_function_calculator(
+                        molecule, iso, state
+                    )
+                    Q, Qvib, dfQrot = parsum.at_noneq(
+                        Tvib,
+                        Trot,
+                        vib_distribution=vib_distribution,
+                        rot_distribution=rot_distribution,
+                        overpopulation=overpopulation,
+                        returnQvibQrot=True,
+                        update_populations=self.misc.export_populations,
+                    )
+
+                    # ... make sure PartitionFunction above is calculated with the same
+                    # ... temperatures, rovibrational distributions and overpopulations
+                    # ... as the populations of active levels (somewhere below)
+                    df.at[idx, "Qvib"] = Qvib
```

Comment:
From here it starts to be a bit complicated, as you'll have to handle multiple partition functions.

You may need it to write another proof of concept to fully understand this case. 
But we may be able to merge the improvements for the equilibrium case already.

Reply:
I Will be setting up a proof of concept as discussed [here.](https://radis-radiation.slack.com/archives/D01PB4L6GV6/p1624953400003900) Marking this comment as resolved here for ease of work on the PR.